scripts/linkml_to_flat_synapse_jsonschema.py

```diff
@@ -0,0 +1,121 @@
+    else:
+        # Add $schema if it doesn't exist
+        data["$schema"] = "https://json-schema.org/draft-07/schema"
+        print("Added $schema field with Draft-07")
+    with open(filepath, "w") as f:
+        json.dump(data, f, indent=2)
+    print(f"Fixed schema version in {filepath}")
+
+def remove_unsupported_fields(filepath):
+    with open(filepath, "r") as f:
+        data = json.load(f)
+    def recursive_clean(obj):
+        if isinstance(obj, dict):
+            for field in ["$defs", "metamodel_version", "version"]:
+                if field in obj:
+                    del obj[field]
+            for v in obj.values():
+                recursive_clean(v)
+        elif isinstance(obj, list):
+            for item in obj:
+                recursive_clean(item)
+    recursive_clean(data)
+    with open(filepath, "w") as f:
+        json.dump(data, f, indent=2)
+    print(f"Cleaned unsupported fields from {filepath}")
+
+def fix_additional_properties(filepath):
+    """Recursively replace boolean additionalProperties with {} in a JSON Schema file."""
+    with open(filepath, "r") as f:
+        data = json.load(f)
+    def recursive_fix(obj):
+        if isinstance(obj, dict):
+            if "additionalProperties" in obj and isinstance(obj["additionalProperties"], bool):
+                obj["additionalProperties"] = {}
+            for v in obj.values():
+                recursive_fix(v)
+        elif isinstance(obj, list):
+            for item in obj:
+                recursive_fix(item)
+    recursive_fix(data)
+    with open(filepath, "w") as f:
+        json.dump(data, f, indent=2)
+    print(f"Fixed additionalProperties in {filepath}")
+
+def main():
+    if len(sys.argv) < 2 or len(sys.argv) > 4:
+        print("Usage: python linkml_to_flat_synapse_jsonschema.py <linkml_yaml> [class_name] [output_filename]")
+        sys.exit(1)
+    linkml_yaml = sys.argv[1]
+    class_name = sys.argv[2] if len(sys.argv) > 2 else ""
+    if len(sys.argv) == 4:
+        output_filename = sys.argv[3]
+    else:
+        base = Path(linkml_yaml).stem
+        output_filename = f"{base}.flat.schema.json"
+    output_dir = Path("JSON_Schemas")
+    output_dir.mkdir(exist_ok=True)
+    output_file = output_dir / output_filename
+    tmp_json = output_file.with_suffix(output_file.suffix + ".tmp.json")
+    # 1. Generate JSON Schema
+    run_gen_json_schema(linkml_yaml, class_name, tmp_json)
+    # 2. Flatten JSON Schema using Python
+    flatten_json_schema(tmp_json, output_file)
+    # 3. Fix schema version to Draft-07
+    fix_schema_version(output_file)
+    # 4. Fix additionalProperties
+    fix_additional_properties(output_file)
+    # 5. Remove unsupported fields
+    remove_unsupported_fields(output_file)
+    # 6. Cleanup
+    Path(tmp_json).unlink(missing_ok=True)
+    print(f"✅ Synapse-compatible flat schema written to {output_file}")
+
+if __name__ == "__main__":
+    main() 
```

Comment:
Lastly, I want to say that I really like how you've laid out the workflow in the `main` function, it's very clear what happens between each of the steps.  If I had to give feedback is that I would read in the jsonschema once, and then write it out at the end. 

Right now you're reading and writing the jsonschema at each step, but that may not be necessary.  One minimal risk is that something can also write to the same jsonschema that is being edited by this code.  Having the jsonschema in memory could mitigate that.

If there are memory overload issues due to the recursive functions, then consider using `tempfile` module to create temporary directory or temporary files.  This way you don't need to do this: `Path(tmp_json).unlink(missing_ok=True)`